src/dialer/queue.js

```diff
@@ -146,10 +147,17 @@ class Queue {
 
   /**
    * Marks the queue as blacklisted. The queue will be immediately aborted.
+   * @returns {void}
    */
   blacklist () {
-    log('blacklisting queue for %s', this.id)
-    this.blackListed = Date.now()
+    this.blackListCount++
+
+    if (this.blackListCount >= 5) {
+      this.blackListed = Infinity
```

Comment:
Infinity seems like a long time :)

I'm thinking of the case where a peer goes offline for a couple of hours and then comes back online.

Reply:
it sure is! Really what I'm trying to do is get a reasonable, increasing backoff. Right now, not including the jitter i just added, it caps at ~5.3 hours (~8.3 hours if you take into account the successive failures it would take to get there). Perhaps this scale isn't sufficient. (5min, 40min, 135min, 320min).

I do think we could scale it back from Infinity though, perhaps a max of 1day would be large enough. This would basically give us the processing of: if I can't dial you over the course of 8 hours, I'll just wait until tomorrow. Each failure after that would also postpone a day. 

Eventually a "dead" node should stop getting emitted as a peer.

Reply:
Having a back-off makes sense 👍 

Maybe we could have another process that removes peers from the blacklist after a certain amount of time so as to limit memory usage. It could be an issue for a node that sees a ton of ephemeral peers (eg a gateway node).

Is there a mechanism by which peers are removed from the blacklist on discovery?

Reply:
What if we had the queue manager go through its queues on a daily/semi-daily(?) interval and had it simply delete any queues with an `Infinity` blacklist? If the peers are discovered again, we could attempt to dial them, and if not we free up the memory of that queue. That might also be a good trigger to remove it from `PeerBook`.

Reply:
That sounds like a reasonable approach. I would favour frequent garbage collection, as it's not a very expensive operation and freeing up memory could be critical for limited memory environments. We might also want to be a config option for a maximum size for the queue.

Reply:
I added an hourly timer to abort and remove queues that have hit a blacklist of `Infinity`. 

> We might also want to be a config option for a maximum size for the queue.

Which queue are you referring to? `_queue`? If so, Ideally we'd avoid setting a max size on this. If we can't get through `_queue` quickly that means our normal node operations are going to suffer. So far my testing locally, has seen `_queue` relatively low and it clears fairly quickly. The cold queue is usually close to max, except right at startup.

Reply:
Sorry I meant the list of dial queues in QueueManager. It just keeps growing in size, right?

Reply:
Yes, there's currently a DialQueue for every peer we've attempted to connect with, unless it's cleaned via the hourly cleanup.

I guess we do need to do some cleanup there. Queues that aren't active and aren't tracking blacklisting shouldn't need to stay around. Ideally I think blacklisting info would be stored as metadata in PeerBook, this would prevent us holding onto queues unnecessarily.

Reply:
I reduced the cleanup time down to 15mins and added some more logic there. If it's not connected, not blacklisted and not queued to be run, we get rid of it. If it matches those other criteria we keep it to avoid frequently recreating the queues.